### PR TITLE
tarreceive: API to convert Docker tar layers

### PR DESCRIPTION
**Pull request by lmars**
This replaces the `docker-receive` app with a much simpler `tarreceive` app which just receives tar image layers from `docker save`, converts them to SquashFS and stores them in the blobstore, leaving the client to construct an image manifest and deploy it.

The main reason for this work is to remove the large dependency on an old version of docker so that we can vendor newer libraries like [cAdvisor](https://github.com/google/cadvisor) without descending into dependency hell 🙈

It also paves the way for building a tool to deploy [OCI images](https://github.com/opencontainers/image-spec) and integrating with [BuildKit](https://github.com/moby/buildkit).

### CODE REVIEW

File: cli/docker.go [hunk 1]
```diff
@@ -371,3 +401,240 @@ func dockerSave(tag string, tw *backup.TarWriter, progress backup.ProgressBar) e
...
+	Layers []string `json:"Layers"`
+}
+
+// DockerConfig is used to read image config from the output of 'docker save'.
+type DockerConfig struct {
+	Config struct {
+		Env        []string `json:"Env"`
+		Cmd        []string `json:"Cmd"`
+		WorkingDir string   `json:"WorkingDir"`
+		Entrypoint []string `json:"Entrypoint"`
```
Comment (titanous) [inline cli/docker.go]: These json tags are redundant as they are the same as the field names.

File: cli/docker.go [hunk 2]
```diff
@@ -371,3 +401,240 @@ func dockerSave(tag string, tw *backup.TarWriter, progress backup.ProgressBar) e
...
+		return fmt.Errorf("error loading docker image config: %s", err)
+	}
+
+	pushRelease := prevRelease.ID
+	if pushRelease == "" {
+		r, err := client.GetAppRelease("controller")
+		if err != nil {
+			return fmt.Errorf("error getting controller release: %s", err)
+		}
+		pushRelease = r.ID
```
Comment (titanous) [inline cli/docker.go]: Is `pushRelease` used?

File: host/cli/bootstrap.go [hunk 3]
```diff
@@ -791,6 +790,17 @@ DELETE FROM volumes WHERE created_at < '%s';`,
 		}
 	}
 
+	if migrateDocker {
+		fmt.Fprintln(os.Stderr, `
+WARN:
+WARN: There are some legacy Docker images in the backup that can no longer
+WARN: be imported to new clusters. You will need to boot a cluster with version
+WARN: less than v20180416.0, import the backup into that, then create a backup
```
Comment (titanous) [inline host/cli/bootstrap.go]: This version will need to be updated.

File: host/cli/bootstrap.go [hunk 4]
```diff
@@ -963,13 +922,12 @@ DELETE FROM volumes WHERE created_at < '%s';`,
...
-			manifestStepMap["docker-receive-wait"],
+			manifestStepMap["tarreceive"],
+			manifestStepMap["tarreceive-route"],
+			manifestStepMap["tarreceive-wait"],
 		}
 		if _, err := steps.RunWithState(ch, state); err != nil {
-			return fmt.Errorf("error deploying docker-receive: %s", err)
+			return fmt.Errorf("error deploying tarreceive: %s", err)
 		}
 	}
```
Comment (titanous) [inline host/cli/bootstrap.go]: We should probably delete the `docker-receive` app?

**Comment by lmars** [comment]
@titanous comments addressed, ready for another review

File: cli/docker.go [hunk 5]
```diff
@@ -10,54 +10,68 @@ import (
...
+	5.50 KB 279.83 KB/s 0s
+	uploading layer a8de0e025d94b33db3542e1e8ce58829144b30c6cd1fff057eec55b1491933c3
+	3.00 KB 153.83 KB/s 0s
+	Docker image deployed, scale it with 'flynn scale app=N'
 `)
 }
 
+// minDockerPushTarVersion is the minimum API version which supports pushing
+// Docker images as tar layers.
+const minDockerPushTarVersion = "v20180905.1"
```
Comment (titanous) [inline cli/docker.go]: I assume this will need updating?

File: host/cli/bootstrap.go [hunk 6]
```diff
@@ -791,6 +790,17 @@ DELETE FROM volumes WHERE created_at < '%s';`,
...
 
+	if migrateDocker {
+		fmt.Fprintln(os.Stderr, `
+WARN:
+WARN: There are some legacy Docker images in the backup that can no longer
+WARN: be imported to new clusters. You will need to boot a cluster with version
+WARN: less than v20180905.1, import the backup into that, then create a backup
```
Comment (titanous) [inline host/cli/bootstrap.go]: Version needs updating

File: host/cli/bootstrap.go [hunk 7]
```diff
@@ -963,16 +922,52 @@ DELETE FROM volumes WHERE created_at < '%s';`,
...
-			manifestStepMap["docker-receive"],
-			manifestStepMap["docker-receive-route"],
-			manifestStepMap["docker-receive-wait"],
+			manifestStepMap["tarreceive"],
+			manifestStepMap["tarreceive-route"],
+			manifestStepMap["tarreceive-wait"],
 		}
 		if _, err := steps.RunWithState(ch, state); err != nil {
-			return fmt.Errorf("error deploying docker-receive: %s", err)
+			return fmt.Errorf("error deploying tarreceive: %s", err)
```
Comment (titanous) [inline host/cli/bootstrap.go]: Should we delete the docker-receive app?

Comment (lmars) [reply host/cli/bootstrap.go]: It gets deleted below (see [here](https://github.com/flynn/flynn/blob/tarreceive/host/cli/bootstrap.go#L934-L969)).

**Pull request merged**